lib/util/progress.coffee

```diff
@@ -1,53 +1,204 @@
-# TODO: move determinate progress bars to the front
-# TODO: hover UI with progress stack, descriptions
+{Emitter}  = require 'atom'
+inkTooltip = require './tooltip'
+
+# Progress Bars
+#
+# This module provides an API for displaying progress bars in Atom. The methods
+# below allow modifing the stack of progress bars, which is represented by
+# corresponding UI elements:
+#    - A "global" progress bar is shown in the status bar. In pratice, this is
+#      the first determinate progress bar in the stack. If there is none, the
+#      first element (which then is indeterminate) will be shown instead. If the
+#      stack is empty, an empty progress bar will be shown instead.
+#    - Hovering over that status bar element will show the complete stack as
+#      an overlay.
+#    - Hovering over the actual progress bar of one such stack element will show
+#      the message correesponding to that progress bar, if there is any.
+#
+# Methods:
+#
+# add(p = {progress: 0})
+#    Create and return a ProgressBar with the initial properties specified by `p`,
+#    which has the following methods available to it:
+#
+#    p.setProgress(prog)
+#        Updates `p`s progress. `prog` is a number between -∞ and 1; if it is
+#        negative, a indeterminate progress bar will be displayed.
```

Comment:
"Determinacy", although we could certainly use a less clumsy terminology if anything springs to mind :)